lib/Doctrine/DBAL/SQLParserUtils.php

```diff
@@ -92,6 +92,7 @@ public static function expandListParameters($query, $params, $types)
 
         if ($isPositional) {
             ksort($params);
+            $types = self::fillMissingTypes($params, $types);
```

Comment:
@kejwmen thanks for fixing this. I believe this problem also exists for named parameters but reveals in a slightly different way. Consider the following test:
```php
array(
    "SELECT * FROM Foo WHERE :foo IN (:bar)",
    array('foo' => 1, 'bar' => array(1, 2)),
    array('bar' => Connection::PARAM_INT_ARRAY),
    "SELECT * FROM Foo WHERE ? IN (?, ?)",
    array(1, 1, 2),
    array(null, \PDO::PARAM_INT, \PDO::PARAM_INT)
),
```
On your branch, it produces:
```
1) Doctrine\Tests\DBAL\SQLParserUtilsTest::testExpandListParameters with data set #10 ('SELECT * FROM Foo WHERE :foo IN (:bar)', array(1, array(1, 2)), array(101), 'SELECT * FROM Foo WHERE ? IN (?, ?)', array(1, 1, 2), array(null, 1, 1))
Types dont match
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
-    0 => null
+    0 => 2
```
Basically, despite the first parameter has an integer value and doesn't have the type specified, it will be bound as a `PDO::PARAM_STR`.

Reply:
If you decide to fix that as well, please make sure it works in the case when parameters have leading colons but types don't and vice versa like in these cases:

https://github.com/doctrine/dbal/blob/09e072dee1457bcdf3cae4be60b18854f8829992/tests/Doctrine/Tests/DBAL/SQLParserUtilsTest.php#L324-L339

Reply:
Thanks for test cases, digging!